### PR TITLE
Add configurable PC2 message

### PR DESCRIPTION
**Pull request by alek-kam**
## What does this PR do?

This PR aims to further improve ROS2 Lidar Component's configurability. It does so by adding message configuration for the ROS2 Lidar component. The message configuration is composed of fields which signify one or more actual PC2 message fields. The presence, order, padding and names of the published fields can now be configured through the ROS2 Lidar Component's editor interface.

_Please describe any testing performed._

This feature was tested under various configurations on a simple scene. __Please note that this is a draft pull request and is not yet ready for review.__

List of features:
- [x] Raycast results handling,
- [x] Editor-side configuration,
- [x] Default value handling,
- [ ] Pointcloud ordering (width, height),
- [x] Removal of Raycaster-side publishing

### CODE REVIEW

File: Gems/ROS2/Code/Include/ROS2/Lidar/RaycastResults.h [hunk 1]
```diff
@@ -52,7 +57,7 @@ namespace ROS2
 
     struct SegmentationIds
     {
-        int32_t m_entityId;
+        AZ::s32 m_entityId;
```
Comment (PawelLiberadzki) [inline Gems/ROS2/Code/Include/ROS2/Lidar/RaycastResults.h:60]: What is the reason for this change?

File: Gems/ROS2/Code/Source/Lidar/Publishing/PointCloudMessageBuilder.cpp [hunk 2]
```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+#include "AzCore/Math/Sfmt.h"
+
+#include <AzCore/std/string/string.h>
```
Comment (PawelLiberadzki) [inline Gems/ROS2/Code/Source/Lidar/Publishing/PointCloudMessageBuilder.cpp]: I would unify brackets / includes.

File: Gems/ROS2/Code/Source/Lidar/Publishing/PointCloudMessageFormat.h [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+        // clang-format off
+        None                    = 0,
+        PositionXYZF32          = 1,
+        IntensityF32            = 1 << 1,
+        TU32                    = 1 << 2, // ?
+        ReflectivityU16         = 1 << 3,
+        RingU16                 = 1 << 4,
+        AmbientU16              = 1 << 5,
+        RangeU32                = 1 << 6,
+        SegmentationData96      = 1 << 7,
```
Comment (PawelLiberadzki) [inline Gems/ROS2/Code/Source/Lidar/Publishing/PointCloudMessageFormat.h]: I would explain this `96` a little, maybe in a comment here.

File: Gems/ROS2/Code/Source/Lidar/Publishing/PointCloudMessageFormat.h [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+        static constexpr Type DefaultValue = 0.0f;
+        static constexpr RaycastResultFlags Provider = RaycastResultFlags::None;
+    };
+
+    template<>
+    struct FieldTraits<FieldFlags::RingU16>
+    {
+        using Type = AZ::u16;
+        static constexpr Type DefaultValue = 0U;
+        static constexpr RaycastResultFlags Provider = RaycastResultFlags::None;
```
Comment (PawelLiberadzki) [inline Gems/ROS2/Code/Source/Lidar/Publishing/PointCloudMessageFormat.h:91]: Ring (channel) can be returned from RGL, but requires additional configuration through API. We can discuss this if it will be necessary.

File: Gems/ROS2/Code/Source/Lidar/Publishing/PointCloudMessageFormat.cpp [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+                        "Description of the types comprising this field.")
+                    ->Attribute(AZ::Edit::Attributes::ContainerCanBeModified, false)
+                    ->Attribute(AZ::Edit::Attributes::Visibility, &FieldFormat::IsRegularField);
+            }
+        }
+    }
+
+    bool FieldFormat::IsRegularField() const
+    {
+        return m_fieldFlag != FieldFlags::Padding8 && m_fieldFlag != FieldFlags::Padding16 && m_fieldFlag != FieldFlags::Padding32;
```
Comment (PawelLiberadzki) [inline Gems/ROS2/Code/Source/Lidar/Publishing/PointCloudMessageFormat.cpp]:
```suggestion
        return !IsPadding();
```
? ;)

**Review by PawelLiberadzki — COMMENTED** [review]
Really nice code, for a draft review I would say the direction is fine and it should work.

File: Gems/ROS2/Code/Include/ROS2/Lidar/RaycastResults.h (hunk 1 above)
Comment (alek-kam) [reply Gems/ROS2/Code/Include/ROS2/Lidar/RaycastResults.h:60]: This is for consistency and expands to int32_t.

File: Gems/ROS2/Code/Source/Lidar/Publishing/PointCloudMessageFormat.h (hunk 4 above)
Comment (alek-kam) [reply Gems/ROS2/Code/Source/Lidar/Publishing/PointCloudMessageFormat.h:91]: I would consider adding this as a raycast result after the rest of the PR is ready (possibly as a separate PR).

**Comment by alek-kam** [comment]
Closing in favor of https://github.com/o3de/o3de-extras/pull/779.
